drivers/at/at.c

```diff
@@ -0,0 +1,238 @@
+                return -1;
+            }
+        }
+        else {
+            return res;
+        }
+    }
+
+    return 0;
+}
+
+void at_send_bytes(at_dev_t *dev, const char *bytes, size_t len)
+{
+    uart_write(dev->uart, (const uint8_t *)bytes, len);
+}
+
+int at_send_cmd(at_dev_t *dev, const char *command, uint32_t timeout)
+{
+    size_t cmdlen = strlen(command);
+
+    uart_write(dev->uart, (const uint8_t *)command, cmdlen);
+    uart_write(dev->uart, (const uint8_t *)AT_EOL, AT_EOL_LEN);
+
+    if (at_expect_bytes(dev, command, timeout)) {
+        return -1;
+    }
+
+    if (at_expect_bytes(dev, AT_EOL "\r\n", timeout)) {
+        return -2;
+    }
+
+    return 0;
+}
+
+void at_drain(at_dev_t *dev)
+{
+    char _tmp[16];
+    int res;
+
+    do {
+        /* consider no character within 10ms "drained" */
+        res = isrpipe_read_timeout(&dev->isrpipe, _tmp, sizeof(_tmp), 10000U);
+    } while (res > 0);
+}
+
+ssize_t at_send_cmd_get_resp(at_dev_t *dev, const char *command,
+                             char *resp_buf, size_t len, uint32_t timeout)
+{
+    ssize_t res = -1;
+
+    at_drain(dev);
+
+    res = at_send_cmd(dev, command, timeout);
+    if (res) {
+        goto out;
+    }
+
+    res = at_readline(dev, resp_buf, len, timeout);
+    if (res == 0) {
+        /* skip possible empty line */
+        res = at_readline(dev, resp_buf, len, timeout);
+    }
+
+out:
+    return res;
+}
+
+ssize_t at_send_cmd_get_lines(at_dev_t *dev, const char *command,
+                              char *resp_buf, size_t len, uint32_t timeout)
+{
+    ssize_t res = -1;
+    size_t bytes_left = len - 1;
+    char *pos = resp_buf;
+
+    memset(resp_buf, '\0', len);
+
+    at_drain(dev);
+
+    res = at_send_cmd(dev, command, timeout);
+    if (res) {
+        goto out;
+    }
+
+    while (1) {
+        res = at_readline(dev, pos, bytes_left, timeout);
+        if (res == 0) {
+            continue;
+        }
+        else if (res > 0) {
+            bytes_left -= res;
+            if ((res == 2) && (strncmp(pos, "OK", 2) == 0)) {
+                res = len - bytes_left;
+                break;
+            }
+            else if ((res == 5) && (strncmp(pos, "ERROR", 5) == 0)) {
+                return -1;
+            }
+            else if (strncmp(pos, "+CME ERROR:", 11) == 0) {
+                return -1;
+            }
+            else if (strncmp(pos, "+CMS ERROR:", 11) == 0) {
+                return -1;
+            }
+            else {
+                pos += res;
+                if (bytes_left) {
+                    *pos++ = '\n';
+                    bytes_left--;
+                }
+                else {
+                    return -1;
+                }
+            }
+        }
+        else {
+            break;
+        }
+    }
+
+out:
+    return res;
+}
+
+int at_send_cmd_wait_prompt(at_dev_t *dev, const char *command, uint32_t timeout)
+{
+    at_drain(dev);
+
+    int res = at_send_cmd(dev, command, timeout);
```

Comment:
I didn't re-test yet, but IIRC, this won't work anymore.
Because in case of the prompt, the "\r\n" is not sent back.

In 'normal' case: we send `command\r\n` and we receive `command\r\n\r\nOK\r\n`. In case of a prompt, if we send `command\r\n` we receive `command\r\n>`. That's why I added a new function instead of reusing `at_send_cmd`.

Reply:
I thought we just send `command\r` (without the `\n`)

Reply:
confusing at least ;)
Anyhow, I reverted to the working version. Let's optimize later.